### ISSUE TITLE
`Truncated` or `truncated` not supported under Zygote backend

### ISSUE THREAD

**Issue by wupeifan** [TuringLang/Turing.jl#1361]
MWE:

``` Julia
using Zygote, ChainRules, Turing, Random
Turing.setadbackend(:zygote)
@model function demo(x)
    rho ~ Truncated(Beta(3, 3), 0.5, 0.99)
    x ~ Normal(3 * rho, 0.1)
end
est_model = demo(2.1)
chain = sample(est_model, NUTS(), 100; progress = true)
```
The above fails with an error message `Can't differentiate foreigncall expression`. 
The error goes away if we comment out the zygote backend line.

**Comment by wupeifan** [comment]
replacing `Truncated` by `truncated` yields the same error.

**Comment by devmotion** [comment]
Copying from Slack:

> I guess I figured out what the issue is in the MWE: in https://github.com/JuliaStats/Distributions.jl/blob/b59a26ad69ff50c04377ecd54ecb198c7b674b74/src/truncate.jl#L65 the cdf of the Beta distribution is evaluated which is a "foreigncall expression" since Distributions uses Rmath for it. DistributionsAD only defines a custom adjoint for the logpdf of the Beta distribution (but not the cdf), so Zygote can't differentiate it. You should be able to fix if you add a custom adjoint for the cdf of the Beta distribution.
>
> I guess one could just use SpecialFunctions instead of Rmath in StatsFuns for defining betacdf and add the adjoint of beta_inc to ChainRules: https://en.wikipedia.org/wiki/Beta_distribution#Cumulative_distribution_function
>
> BTW you should use truncated instead of Truncated  in general. truncated might use some specialized truncated distribution type whereas Truncated always creates an instance of type Truncated.

**Comment by wupeifan** [comment]
Thanks for digging into this!
I guess there's no easy fix for this one then. It would be great if there's a list to show which distributions we can `Truncate` under Zygote -- basically which distributions have cdf explicitly defined in Julia.

**Comment by wupeifan** [comment]
Hmm, then I think we can close this.
tl;dr for anyone who check this thread later: if the `cdf` of the distribution is defined in native Julia in `Distributions`, then the MWE should work (e.g. Normal distribution). Otherwise, it will fail (e.g. Beta, Gamma distributions)
